text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,146 @@
+
+Types of connection details will be defined by an enum with only even number values. Odd number values will be reserved for potential third-party protocols (RFC 39).
+
+List of types:
+```c
+enum obs_service_connect_info {
+	OBS_SERVICE_CONNECT_INFO_SERVER_URL = 0,
+	OBS_SERVICE_CONNECT_INFO_STREAM_KEY = 2,
+	OBS_SERVICE_CONNECT_INFO_USERNAME = 4,
+	OBS_SERVICE_CONNECT_INFO_PASSWORD = 6,
+	OBS_SERVICE_CONNECT_INFO_STREAM_ID = 8,
+	OBS_SERVICE_CONNECT_INFO_ENCRYPT_PASSPHRASE = 10,
+};
+```
+
+Adding to `obs_service_info`:
+- `const char *(*get_connect_info)(void *data, uint32_t type)` where `type` is one of the value of the list above indicating which info we want and return `NULL` if it doesn't have it.
+- `bool (*can_try_to_connect)(void *data)`, since protocols and services do not always require the same information. This function allows the service to return if it can connect. Return true if the the service has all it needs to connect.
+
+Adding these functions to the Services API:
+- `const char *obs_service_get_connect_info(const obs_service_t *service, uint32_t type)`: return the connection information related to the given type (list above).
+- `bool obs_service_can_try_to_connect(const obs_service_t *service)`: return if the service can try to connect.
+  - return `bool (*can_try_to_connect)(void *data)` result.
+  - return true if `bool (*can_try_to_connect)(void *data)` is not implemented in the service.
+  - return false if the service object is `NULL`.
+
+`obs_service_get_url()`, `obs_service_get_key()`, `obs_service_get_username()`, and `obs_service_get_password()` will be deprecated in favor of `obs_service_get_connect_info()`.
+
+In the future, if we want to support a protocol that doesn't use a server URL, this scenario is covered by this design.
+
+### About `rtmp-services`
+
+`rtmp-services` is a plugin that provides two services `rtmp_common` and `rtmp_custom`. The use of "rtmp" in these three terms no longer means that it only supports RTMP, it was and is kept to avoid breakage.
+
+- In `rtmp_common`, services must at least support H264 as video codec and AAC or Opus as audio codec. This is a limitation required by the simple output mode.
+- The service `rtmp_custom` will depend on protocol prefixes from outputs types to detect which protocol is in use.
+
+#### About service JSON file
+
+- For `services` objects, a `"protocol"` field will be added. This will be used to indicate the protocol for services in situation where:
+  - The protocol can not be deduced from server's URL prefix.
+  - The service is RTMP but has an HTTP URL as its server URL and it has custom code to request the true server URL through an API in the plugin. In this case the protocol will be enforced to the right protocol.
+- Services that use a protocol that is not registered will not be shown. e.g. OBS Studio without RTMPS support will not show services and servers that rely on RTMPS.
```

Comment:
Instead of starting a full sentence with "e.g.", I would just say "For example, ".